src/typechecker.rs

```diff
@@ -3246,46 +3250,118 @@ pub fn typecheck_statement(
+                                    Some(JaktError::TypecheckError(
+                                        "unpacked tuples must match in size".to_string(),
+                                        init.span(),
+                                    )),
+                                );
+                            }
 
-            let checked_var_decl = CheckedVarDecl {
-                name: var_decl.name.clone(),
-                type_id: checked_type_id,
-                span: var_decl.span,
-                mutable: var_decl.mutable,
-                visibility: var_decl.visibility.clone(),
-            };
+                            for (idx, name) in inner_names.iter().enumerate() {
+                                match &project.types[inner[idx]] {
+                                    Type::GenericInstance(struct_id, _)
+                                        if *struct_id == weakptr_struct_id =>
+                                    {
+                                        if !name.mutable {
+                                            error = error.or(Some(JaktError::TypecheckError(
+                                                "Weak reference must be mutable".into(),
+                                                var_decl.span,
+                                            )));
+                                        }
+                                    }
+                                    _ => {}
+                                }
+                                let variable = CheckedVariable {
+                                    name: name.name.clone(),
+                                    type_id: inner[idx],
+                                    mutable: name.mutable,
+                                    visibility: var_decl.visibility.clone(),
+                                    definition_span: var_decl.span, // keep track of the individual variables spans
+                                };
 
-            let err = project
-                .add_var_to_scope(
-                    scope_id,
-                    CheckedVariable {
-                        name: checked_var_decl.name.clone(),
-                        type_id: checked_var_decl.type_id,
-                        mutable: checked_var_decl.mutable,
-                        visibility: checked_var_decl.visibility.clone(),
-                        definition_span: checked_var_decl.span,
-                    },
-                    checked_var_decl.span,
-                )
-                .err();
-            error = error.or(err);
+                                decls.push(variable);
+                            }
+                        }
 
-            (
-                CheckedStatement::VarDecl(checked_var_decl, checked_expression),
-                error,
-            )
+                        _ => {
+                            return (
+                                CheckedStatement::Garbage,
```

Comment:
This won't work with generic functions, it'll appear to have successfully typechecked, but it will generate no code in the function, leading to errors
Please either add an error type that's not ignored pre-generic instantiation, or produce _something_ that makes sense given the known types and data.

Reply:
Do you have an example that I could try?